src/Symfony/Component/Messenger/Bridge/GooglePubSub/Transport/Connection.php

```diff
@@ -0,0 +1,188 @@
+use Google\Cloud\PubSub\PubSubClient;
+use Google\Cloud\PubSub\Subscription;
+use Google\Cloud\PubSub\Topic;
+use Symfony\Component\Messenger\Exception\InvalidArgumentException;
+use Symfony\Component\Messenger\Exception\TransportException;
+
+/**
+ * @author Richard Henkenjohann <richardhenkenjohann@googlemail.com>
+ *
+ * @internal
+ */
+class Connection
+{
+    private const DEFAULT_OPTIONS = [
+        // PubSub client
+        'apiEndpoint' => null,
+        'projectId' => null,
+        'keyFile' => null,
+        'keyFilePath' => null,
+        'requestTimeout' => null,
+        'retries' => 3,
+        'scopes' => null,
+        'quotaProject' => null,
+
+        // PubSub topic
+        'topic' => [
+            'name' => null
+        ],
+
+        // PubSub subscription
+        'subscription' => [
+            'name' => null
+        ],
+
+        // Internal
+        'auto_setup' => true,
+    ];
+
+    private $configuration;
+    private $client;
+
+    public function __construct(array $configuration, PubSubClient $client = null)
+    {
+        $this->configuration = array_replace_recursive(self::DEFAULT_OPTIONS, $configuration);
+        $this->client = $client ?? new PubSubClient([]);
+    }
+
+    /**
+     * Creates a connection based on the DSN and options.
+     *
+     * Available options:
+     *
+     * * apiEndpoint: The hostname with optional port to use in place of the default service endpoint
+     * * projectId: The project ID from the Google Developer's Console
+     * * keyFile: The contents of the service account credentials.json file retrieved from the Google Developer's Console.
+     * * keyFilePath: The full path to your service account credentials .json file retrieved from the Google Developers Console.
+     * * requestTimeout: Seconds to wait before timing out the request. (Defaults: `0` with REST and `60` with gRPC)
+     * * retries: Number of retries for a failed request. (Default: `3`)
+     * * scopes: Scopes to be used for the request
+     * * quotaProject: Specifies a user project to bill for access charges associated with the request
+     * * topic: The options for the topic to publish messages to
+     * * subscription: The options of the subscription to receive messages from
+     * * auto_setup: Whether the queue should be created automatically during send / get (Default: true)
+     */
+    public static function fromDsn(string $dsn, array $options = []): self
+    {
+        if (false === $parsedUrl = parse_url($dsn)) {
+            throw new InvalidArgumentException(sprintf('The given Google Pub/Sub DSN "%s" is invalid.', $dsn));
+        }
+
+        $query = [];
+        if (isset($parsedUrl['query'])) {
+            parse_str($parsedUrl['query'], $query);
+        }
+
+        // Check for unsupported keys in options
+        $optionsExtraKeys = array_diff(array_keys($options), array_keys(self::DEFAULT_OPTIONS));
+        if (0 < \count($optionsExtraKeys)) {
```

Comment:
```suggestion
        if (!empty($optionsExtraKeys)) {
```

Reply:
(we try to avoid using empty() when possible)

Reply:
I see, all right then.